gax/src/main/java/com/google/api/gax/batching/BlockingSemaphore.java

```diff
@@ -34,40 +34,90 @@
 
 /** A {@link Semaphore64} that blocks until permits become available. */
 class BlockingSemaphore implements Semaphore64 {
-  private long currentPermits;
+  private long availablePermits;
+  private long limit;
 
   private static void checkNotNegative(long l) {
     Preconditions.checkArgument(l >= 0, "negative permits not allowed: %s", l);
   }
 
   BlockingSemaphore(long permits) {
     checkNotNegative(permits);
-    this.currentPermits = permits;
+    this.availablePermits = permits;
+    this.limit = permits;
   }
 
+  @Override
   public synchronized void release(long permits) {
     checkNotNegative(permits);
-
-    currentPermits += permits;
+    // TODO: throw exceptions when the permits overflow
+    availablePermits = Math.min(availablePermits + permits, limit);
     notifyAll();
   }
 
+  @Override
   public synchronized boolean acquire(long permits) {
     checkNotNegative(permits);
 
     boolean interrupted = false;
-    while (currentPermits < permits) {
+    while (availablePermits < permits) {
+      try {
+        wait();
+      } catch (InterruptedException e) {
+        interrupted = true;
+      }
+    }
+    // TODO: if thread is interrupted, we should not grant the permits
+    availablePermits -= permits;
+
+    if (interrupted) {
+      Thread.currentThread().interrupt();
+    }
+    return true;
+  }
+
+  @Override
+  public synchronized boolean acquirePartial(long permits) {
+    checkNotNegative(permits);
+
+    boolean interrupted = false;
+    // To allow individual oversized requests to be sent, clamp the requested permits to the maximum
+    // limit. This will allow individual large requests to be sent. Please note that this behavior
+    // will result in availablePermits going negative.
+    while (availablePermits < Math.min(limit, permits)) {
       try {
         wait();
       } catch (InterruptedException e) {
         interrupted = true;
       }
     }
-    currentPermits -= permits;
 
     if (interrupted) {
       Thread.currentThread().interrupt();
     }
+
+    availablePermits -= permits;
     return true;
   }
+
+  @Override
+  public synchronized void increasePermitLimit(long permits) {
+    checkNotNegative(permits);
+    availablePermits += permits;
+    limit += permits;
+    notifyAll();
+  }
+
+  @Override
+  public synchronized void reducePermitLimit(long reduction) {
+    checkNotNegative(reduction);
+    Preconditions.checkState(limit - reduction > 0, "permit limit underflow");
+    availablePermits -= reduction;
+    limit -= reduction;
```

Comment:
I see `increasePermitLimit` calls notifyAll(), but reduce does not. Looks suspicious, as it is a reverse operation to increase (so I would assume it should follow the same principles).

Reply:
It should be ok here, because the wait() in other threads are waiting for more permits to become available, while this method removes permits. So even if we call notifyAll() here, the threads are likely to go back to sleep again :). It's similar to how we don't call notifyAll() in reserve().